src/modules/flow/iio/nodes.c

```diff
@@ -25,57 +25,101 @@
 
 #include <sol-iio.h>
 
-struct iio_direction_vector_data {
+struct iio_device_config {
     struct sol_iio_config config;
-    struct sol_direction_vector scale;
-    struct sol_direction_vector offset;
     struct sol_drange_spec out_range;
     struct sol_iio_device *device;
-    struct sol_iio_channel *channel_x;
-    struct sol_iio_channel *channel_y;
-    struct sol_iio_channel *channel_z;
     bool buffer_enabled : 1;
     bool use_device_default_scale : 1;
     bool use_device_default_offset : 1;
 };
 
-struct color_data {
-    struct sol_iio_config config;
+/* Make sure the iio_device_config is the first element */
+struct iio_double_data {
+    struct iio_device_config iio_base;
     double scale;
     double offset;
-    struct sol_drange_spec out_range;
-    struct sol_iio_device *device;
+    struct sol_iio_channel *channel_val;
+};
+
+struct iio_direction_vector_data {
+    struct iio_device_config iio_base;
+    struct sol_direction_vector scale;
+    struct sol_direction_vector offset;
+    struct sol_iio_channel *channel_x;
+    struct sol_iio_channel *channel_y;
+    struct sol_iio_channel *channel_z;
+};
+
+struct iio_color_data {
+    struct iio_device_config iio_base;
+    double scale_red;
+    double offset_red;
+    double scale_green;
+    double offset_green;
+    double scale_blue;
+    double offset_blue;
     struct sol_iio_channel *channel_red;
     struct sol_iio_channel *channel_green;
     struct sol_iio_channel *channel_blue;
-    bool buffer_enabled : 1;
-    bool use_device_default_scale : 1;
-    bool use_device_default_offset : 1;
 };
 
-struct iio_double_data {
-    struct sol_iio_config config;
-    struct sol_drange_spec out_range;
-    double scale;
-    double offset;
-    struct sol_iio_device *device;
-    struct sol_iio_channel *channel_val;
-    bool buffer_enabled : 1;
-    bool use_device_default_scale : 1;
-    bool use_device_default_offset : 1;
+struct iio_node_type {
+    struct sol_flow_node_type base;
+    uint16_t out_port;
+    void (*reader_cb)(void *data, struct sol_iio_device *device);
 };
 
 static void
-reader_cb(void *data, struct sol_iio_device *device)
+iio_common_close(struct sol_flow_node *node, void *data)
+{
+    struct iio_device_config *mdata = data;
+
+    free((char *)mdata->config.trigger_name);
+    if (mdata->device)
+        sol_iio_close(mdata->device);
+}
+
+static int
+iio_common_tick(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id, const struct sol_flow_packet *packet)
+{
+    static const char *errmsg = "Could not read channel values";
+    struct iio_device_config *mdata = data;
+    struct iio_node_type *type;
+
+    type = (struct iio_node_type *)sol_flow_node_get_type(node);
+
+    if (mdata->buffer_enabled) {
+        if (sol_iio_device_trigger(mdata->device) < 0)
+            goto error;
+    } else {
+        SOL_NULL_CHECK_GOTO(type->reader_cb, error);
```

Comment:
I don't think you need this check: it's you that controls this `type->reader_cb`, so it should never be `NULL`.